### PR TITLE
Support page messaging

### PR DESCRIPTION
**Pull request by godd9170**
I've added an optional `extensionId` to the `Store` constructor. I use the extensionId in the `connect` and `sendMessage` functions, and listen for `onMessageExternal` and `onConnectExternal` in the `background.js`. fixes #49

### CODE REVIEW

File: src/store/Store.js [hunk 1]
```diff
@@ -10,12 +10,12 @@ class Store {
...
    */
-  constructor({portName, state = {}}) {
+  constructor({portName, state = {}, extensionId = ''}) {
     if (!portName) {
       throw new Error('portName is required in options');
     }
-
-    this.port = chrome.runtime.connect({name: portName});
+    this.extensionId = extensionId; //keep the extensionId as an instance variable
+    this.port = chrome.runtime.connect(this.extensionId, {name: portName});
```
Comment (tshaddix) [inline src/store/Store.js]: @godd9170 Have you verified this works when it is passed in as an empty string? We should test on the examples repo :)

File: src/wrap-store/wrapStore.js [hunk 2]
```diff
@@ -40,23 +40,23 @@ export default (store, {
...
+  };
 
   /**
-   * Setup extended connection for state updates
-   */
-  chrome.runtime.onConnect.addListener((port) => {
+  * Setup for state updates
+  */
+  const connectState = (port) => {
+    console.log('connected on port: ', port);
```
Comment (tshaddix) [inline src/wrap-store/wrapStore.js]: Can we remove this console log?

File: src/store/Store.js [hunk 3]
```diff
@@ -64,7 +66,9 @@ class Store {
    */
   dispatch(data) {
     return new Promise((resolve, reject) => {
-      chrome.runtime.sendMessage({
+      chrome.runtime.sendMessage(
+        this.extensionId,
```
Comment (tshaddix) [inline src/store/Store.js]: Have we verified this works with an empty extension id?

File: src/wrap-store/wrapStore.js [hunk 4]
```diff
@@ -80,5 +80,34 @@ export default (store, {
 
     // send initial state
     sendState();
+  };
+
+  /**
+   * Setup action handler
+   */
+  chrome.runtime.onMessage.addListener((request, sender, sendResponse) => {
```
Comment (tshaddix) [inline src/wrap-store/wrapStore.js]: Can we simplify this as `chrome.runtime.onMessage.addListener(dispatchResponse)`

File: src/wrap-store/wrapStore.js [hunk 5]
```diff
@@ -80,5 +80,34 @@ export default (store, {
...
+    return dispatchResponse(request, sender, sendResponse);
   });
+
+  /**
+   * Setup external action handler
+   */
+  chrome.runtime.onMessageExternal.addListener((request, sender, sendResponse) => {
+    return dispatchResponse(request, sender, sendResponse);
```
Comment (tshaddix) [inline src/wrap-store/wrapStore.js]: Same here

File: src/wrap-store/wrapStore.js [hunk 6]
```diff
@@ -80,5 +80,34 @@ export default (store, {
...
+  });
+
+  /**
+   * Setup extended connection
+   */
+  chrome.runtime.onConnect.addListener((port) => {
+    return connectState(port);
```
Comment (tshaddix) [inline src/wrap-store/wrapStore.js]: Can we simplify this to `chrome.runtime.onConnect.addListener(connectState)`

File: src/wrap-store/wrapStore.js [hunk 7]
```diff
@@ -80,5 +80,34 @@ export default (store, {
...
+  });
+
+  /**
+   * Setup extended external connection 
+   */
+  chrome.runtime.onConnectExternal.addListener((port) => {
+    return connectState(port);
```
Comment (tshaddix) [inline src/wrap-store/wrapStore.js]: Same here

**Comment by vhmth** [comment]
@tshaddix you mean to link to a PR? I think you linked to a file :-P

**Comment by tshaddix** [comment]
@vhmth Oh sorry yes I meant to link to a file. I was showing where the unit test needed to be changed. I understand the confusion by saying "this PR" :P

File: src/wrap-store/wrapStore.js (hunk 7 above)
Comment (godd9170) [inline src/wrap-store/wrapStore.js]: 👍

File: src/wrap-store/wrapStore.js (hunk 6 above)
Comment (godd9170) [inline src/wrap-store/wrapStore.js]: 👍

File: src/wrap-store/wrapStore.js (hunk 5 above)
Comment (godd9170) [inline src/wrap-store/wrapStore.js]: 👍

File: src/wrap-store/wrapStore.js (hunk 4 above)
Comment (godd9170) [inline src/wrap-store/wrapStore.js]: Seems pretty reasonable!

File: src/wrap-store/wrapStore.js (hunk 2 above)
Comment (godd9170) [inline src/wrap-store/wrapStore.js]: Did I put that one in? Sorry I thought that was from the original. Sure can.

File: src/store/Store.js (hunk 3 above)
Comment (godd9170) [inline src/store/Store.js]: 👍

File: src/store/Store.js (hunk 1 above)
Comment (godd9170) [inline src/store/Store.js]: I have yes, I tried it with your original `clicker-key` example happening in the `content.js`.

**Comment by godd9170** [comment]
@tshaddix glad you like it, apologies about breaking all your tests, I was too tired to do them last night! I'll make those changes later tonight, do you also want me to take a stab at re-writing the unit tests? I'm afraid to admit I've never done any javascript unit testing, but keen to learn.

**Comment by tshaddix** [comment]
@godd9170 Absolutely no worries! A lot of good additions in here :) 

It's up to you! I think it's a good way to round off this PR and the problem space if pretty small, so it might be good place for you to try it out!

If not, I'll be happy to fix them up. Just let me know and I'll help where I can!

**Comment by godd9170** [comment]
Ok, changes applied and tests fixed. I think we should probably mock up the `sendMessageExternal` stuff, I just couldn't figure out how to do it.

File: src/store/Store.js (hunk 1 above)
Comment (tshaddix) [inline src/store/Store.js]: Okay, great. I will do it on my end as well so we can get two confirmations involved 👍

**Comment by tshaddix** [comment]
Awesome work fixing up the tests! Going to run this one more time on my end  against the examples and then I think we can merge.

**Comment by tshaddix** [comment]
@godd9170 Your changes have been released in `v1.2.0`! Thank you!

**Pull request merged**